### ISSUE TITLE
Form data is lost upon tab/app switch

### ISSUE THREAD

**Issue by ruedigerkupper** [ubports/morph-browser#263]
This turns out to be a real show-stopper:
I start to fill in a form in a web page (e.g., a forum post, account data, account details in online banking, etc.) I need to switch tabs inside Morph or temporarily switch to another app to look up some data (like bank account number, etc.). When I switch back, the page reloads and everything I had already entered is lost.

**Comment by ruedigerkupper** [comment]
This is Morph on a Nexus 5, RC channel.

**Comment by ruedigerkupper** [comment]
Expected behavior: Form data is retained between tab/app switches.
(Sorry, cannot edit original description, the edit button from the dots menu is not working...)

**Comment by balcy** [comment]
yes I have a BQ Aquaris 4.5 and see the same problems.

I would suggest deactivating the Tab unloading in morph-browser. (this is currently done for background tabs, if the free memory is below a certain level) by removing the block https://github.com/ubports/morph-browser/blob/7b4e4f8be52f4ea25022b4e3e1719e9b0377fb1f/src/app/webbrowser/morph-browser.qml#L454 from morph-browser-qml
With that I could successfully use several tabs in morph in switching between them without reload.
[without switching to other apps]

If I switch between RSS reader and morph-browser, the one I switch too will now usually "blurry", then a black screen, and then starts up again, or they crash.

QtWebEngine (and maybe other processes, too) seem to use much more memory than before with the new system. I really don't know the cause.

similar issue (reopened) : https://github.com/ubports/ubuntu-touch/issues/1265

**Comment by balcy** [comment]
not sure if it is a good decision to remove the unloading feature.
advantage: if you stay within morph-browser, you can switch between several tabs
disadvantage: switching between morph and another app (especially webapp) resets morph or the webapp or even crashes.
The real cause for the "memory hungryness" is still unknown to me.

**Comment by ruedigerkupper** [comment]
Common use case (at least with me):
- try to log into a site that requires authorization.
- the site requires my login name and password
- I type my login name from memory, but …
-> … trying to be safe, I do not use common passwords across sites. Instead I use a password vault for storing my credentials. There are several applications for UBPorts and many web apps for doing so. In my case, I use the Passwords app with my Nextcloud instance. In any case whatsoever, I need to switch applications to retrieve my password for the site.
-> as I return, I find that my already typed login name has vanished from the form
- I mourne and type the name again, then paste my password from my password vault to the respective field.
- Next, the site asks for my two-factor authorization code. To get that, I need to switch to the TBOTP-app for UBPorts (there are currently two to my knowledge, both work very well). I switch there, copy my one-time-password and switch back to Morph.
-> Damn! All of my form data is lost! At this time I am really f******, because I do have my one-time-code in the clipboard, but not my login password any longer. I can retype my login name from memory (for the third time!) and paste my one-time-code, but then I need to switch applications again to get my password from my password vault. No to mention that my one-time-code will be already invalid by the time I get back to Morph for pasting my password. But who cares, both my login name and my one-time-code are lost again when I switch back.

… this keeps on indefinitely. I can only have one code in the clipboard at a time, either my password or the one time code for two factor authorization. But each time I copy one of them from the respective app, the other is lost from Morph's form. Usually I just give up logging into the site from my phone and wait until I am home to use my laptop.

In short: Morph's "unloading feature" renders it impossible to login (safely) into web sites requiring two-factor authentication.

**Comment by balcy** [comment]
yes agreed to all of the points, it's hard to use the phone currently on devel channel like that, and I would really find the cause for that. For me it would clearly be a blocker for OTA-12.

> In short: Morph's "unloading feature" renders it impossible to login (safely) into web sites requiring two-factor authentication.

With this I disagree: Within morph only what happens when switching between tabs can be controlled, as soon as you switch to another app, there is nothing morph can do about it being unloaded / killed / ...

[edit]
Only a part of the issue "Form data is lost upon **tab** switch" could be solved be removing the block, but not the general problem. (too little memory, morph unloaded )

You can deactive the unloading of background tabs on your phone by removing the part
 ```property var memoryPressureMonitor: Connections { ```
 (lines 454 to 497) but I would guess switching from the two factor auth app to morph has then still problems (on my phone I've removed the block already, and still cannot use RSS reader and morph-browser at the same time)

My summary would be this:
- there has been no change of QtWebEngine recently (so still same chromium)
- no major changes to morph-browser that would explain it uses much more memory
- the operating system of the dev channel now has a different behavior here

**Comment by balcy** [comment]
on my phone the problem went away by reinstalling the dev image.

But generally, since the new unity8 seems to "defend" itself better against crashes, it might be possible to lower the memory threshold or remove the unloading of tabs. As shown in this issue the browser cannot really decide which tab (even if it is in background) is important to the user, it might contain an information that needs to be copied to another tab.

**Comment by balcy** [comment]
@ruedigerkupper if you use / have used uadblock, I think it might help to switch to another channel (e.g. stable) via system settings (install the ~ 400 MB image), and then after flashing stable back to RC, without activating uadblock again.
At least for my device that is the most likely cause. Just deactivating uadblock did not help for me.

@UniversalSuperBox should we still lower the memory limit ? https://github.com/ubports/morph-browser/issues/99#issuecomment-442171161

**Comment by ruedigerkupper** [comment]
Ah, thanks, I'll try that.

**Comment by UniversalSuperBox** [comment]
@balcy, do you mean increasing the amount of memory that the browser may use (decreasing the low-memory threshold) or decreasing the amount of memory it may use (increasing the low-memory threshold)?

**Comment by balcy** [comment]
I meant `var lowMemoryThresholdPercent = 0.2;` -> `var lowMemoryThresholdPercent = 0.1;` for example

**Comment by UniversalSuperBox** [comment]
Since we have still not added the oom score back to Upstart jobs, that is probably not safe to do. It would cause Unity to be killed more often, taking the entire user session with it.

**Comment by ruedigerkupper** [comment]
While this issue was virtualy gone during the last year, it appears that it's becoming acute again since a few days.
